server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java

```diff
@@ -235,6 +235,16 @@ public boolean cancelForShard(ShardId shardId, String reason) {
         return cancelled;
     }
 
+    /**
+     * check if a shard is currently replicating
+     *
+     * @param shardId      shardId for which to check if replicating
+     * @return true if shard is currently replicating
+     */
+    public boolean isShardReplicating(ShardId shardId) {
+        return onGoingTargetEvents.values().stream().anyMatch(t -> t.indexShard.shardId().equals(shardId));
```

Comment:
Nit-pick - this is a problem to be solved down the line:

`isShardReplicating` seems like a frequent check, and constantly flattening the map to a stream of values may end up being performance-intensive. We should consider if we could make the key of the map the `shardId` to speed up this check.

Reply:
Yes we can do that. But we have to change it in multiple places, so thinking of doing that in different PR